examples/lda.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import absolute_import, division, print_function
+
+import argparse
+import logging
+
+import torch
+from torch import nn
+from torch.distributions import constraints
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro.infer import SVI, JitTraceEnum_ELBO, TraceEnum_ELBO
+from pyro.optim import Adam
+
+logging.basicConfig(format='%(relativeCreated) 9d %(message)s', level=logging.INFO)
+
+
+# This is a fully generative model of a batch of documents.
+# data is a [num_words_per_doc, num_documents] shaped array of word ids
+# (specifically it is not a histogram). We assume in this simple example
+# that all documents have the same number of words.
+@poutine.broadcast
+def model(args, data=None):
+    # Globals.
+    with pyro.iarange("topics", args.num_topics):
+        topic_weights = pyro.sample("topic_weights", dist.Gamma(1. / args.num_topics, 1.))
+        topic_words = pyro.sample("topic_words",
+                                  dist.Dirichlet(torch.ones(args.num_words) / args.num_words))
+
+    # Locals.
+    with pyro.iarange("documents", args.num_docs) as ind:
+        if data is not None:
+            assert data.shape == (args.num_words_per_doc, args.num_docs)
+            data = data[:, ind]
+        doc_topics = pyro.sample("doc_topics", dist.Dirichlet(topic_weights))
+        with pyro.iarange("words", args.num_words_per_doc):
+            # The word_topics variable is marginalized out during inference,
+            # achieved by specifying infer={"enumerate": "parallel"} and using
+            # TraceEnum_ELBO for inference. Thus we can ignore this variable in
+            # the guide.
+            word_topics = pyro.sample("word_topics", dist.Categorical(doc_topics),
+                                      infer={"enumerate": "parallel"})
+            data = pyro.sample("doc_words", dist.Categorical(topic_words[word_topics]),
+                               obs=data)
+
+    return topic_weights, topic_words, data
+
+
+# We will use amortized inference of the local topic variables, achieved by a
+# multi-layer perceptron. We'll wrap the guide in an nn.Module.
+class Guide(nn.Module):
+    def __init__(self, args):
+        super(Guide, self).__init__()
+        layer_sizes = ([args.num_words] +
+                       [int(s) for s in args.layer_sizes.split('-')] +
+                       [args.num_topics])
+        logging.info('Creating MLP with sizes {}'.format(layer_sizes))
+        layers = []
+        for in_size, out_size in zip(layer_sizes, layer_sizes[1:]):
+            layer = nn.Linear(in_size, out_size)
+            layer.weight.data.normal_(0, 0.001)
+            layer.bias.data.normal_(0, 0.001)
+            layers.append(layer)
+            layers.append(nn.Sigmoid())
+        self.local_guide = nn.Sequential(*layers)
+
+    @poutine.broadcast
+    def __call__(self, args, data):
```

Comment:
i _think_ it's more standard to define a `forward()` function, which `nn.Module`'s invoke on `__call__`

Reply:
Thanks. I ended up following @eb8680's suggested which obviates the need for `.forward()`.